File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations
def solution(k, dungeons):
    answer = -1
    
    for i in permutations(dungeons,len(dungeons)):
        temp=k
        cnt=0
        for need,cost in i:
            if temp>=need:
                temp-=cost
                cnt+=1
        answer=max(cnt,answer)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    private int answer = 0;
    
    public int solution(int k, int[][] dungeons) {
        boolean[] visited=new boolean[dungeons.length];
        dfs(k,dungeons,visited,0);
        return answer;
    }
    
    private void dfs(int k,int[][] dungeons,boolean[] visited,int count){
        answer=Math.max(count,answer);
        for(int i=0;i<dungeons.length;i++){
            if(!visited[i]&&k>=dungeons[i][0]){
                visited[i]=true;
                dfs(k-dungeons[i][1],dungeons,visited,count+1);
                visited[i]=false;
            }
        }
    }
}